algoliasearch/account_client.go

```diff
@@ -0,0 +1,123 @@
+package algoliasearch
+
+import "fmt"
+
+type accountClient struct{}
+
+func NewAccountClient() AccountClient {
+	return &accountClient{}
+
+}
+
+func (a *accountClient) CopyIndex(src, dst Index) (err error) {
+	if src.GetAppID() == dst.GetAppID() {
+		return SameAppIDErr
+	}
+
+	if _, err := dst.GetSettings(); err == nil {
+		return IndexAlreadyExistsErr
+	}
+
+	// Copy objects
+	{
+		it, err := src.BrowseAll(nil)
+		if err != nil {
+			return fmt.Errorf("cannot browse source index records: %v", err)
+		}
+
+		var records []Object
+
+		for {
+			res, err := it.Next()
+			if err != nil {
+				if err == NoMoreHitsErr {
+					break
+				} else {
+					return fmt.Errorf("error while iterating source index records: %v", err)
+				}
+			}
+			records = append(records, Object(res))
```

Comment:
Unless I am missing something here, there is an algorithm problem in your code. You are not leveraging the fact that `it` is an iterator -  and your records variable here may cause a `memory exhausted`. You should try todo something similar to:

```python
        # Copy objects
        responses = []
        batch = []
        batch_size = 1000
        count = 0
        for obj in source_index.browse_all():
            batch.append(obj)
            count += 1

            if count == batch_size:
                response = destination_index.save_objects(batch, request_options)
                responses.append(response)
                batch = []
                count = 0

        if batch:
            response = destination_index.save_objects(batch, request_options)
            responses.append(response)

        return responses
```

Reply:
This is an interesting point. I thought that I would collect all the records in memory, which indeed may lead to a high memory footprint, depending on the index' total number of records, to leverage the `ReplaceAllObjects` that performs the automatic batching (1k objects at a time). As it's an helper method, I wanted to keep it simple and not introduce too much complexity, hence my use of the `replaceAll*` methods with the total number of objects. Do you think it's still unreasonable? cc @Ant-hem what are you thoughts on this?

Reply:
IMHO all the features we provide should be scalable.

Reply:
@aseure  I think it's good idea to split the records. I have done this in Java and C#V6 for three main reasons:  memory, network and scaling. It's also great because our users don't have to worry about their lists size.

IMHO, next gen client should implement this behavior by default for `save*` methods.

Reply:
As discussed with all of you, we will rather *not* use the `ReplaceAllObjects` here (to avoid useless temporary index) and we'll send the batch every time it reaches the batch size to avoid consuming too much memory. Thank you all 👍